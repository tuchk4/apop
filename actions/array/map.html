
<!DOCTYPE HTML>
<html lang="" >
    <head>
        <meta charset="UTF-8">
        <meta content="text/html; charset=utf-8" http-equiv="Content-Type">
        <title>map Â· GitBook</title>
        <meta http-equiv="X-UA-Compatible" content="IE=edge" />
        <meta name="description" content="">
        <meta name="generator" content="GitBook 3.2.2">
        
        
        
    
    <link rel="stylesheet" href="../../gitbook/style.css">

    
            
                
                <link rel="stylesheet" href="../../gitbook/gitbook-plugin-timeline/timeline.css">
                
            
                
                <link rel="stylesheet" href="../../gitbook/gitbook-plugin-codetabs/codetabs.css">
                
            
                
                <link rel="stylesheet" href="../../gitbook/gitbook-plugin-disqus/plugin.css">
                
            
                
                <link rel="stylesheet" href="../../gitbook/gitbook-plugin-highlight/website.css">
                
            
                
                <link rel="stylesheet" href="../../gitbook/gitbook-plugin-search/search.css">
                
            
                
                <link rel="stylesheet" href="../../gitbook/gitbook-plugin-fontsettings/website.css">
                
            
                
                <link rel="stylesheet" href="../../gitbook/gitbook-plugin-theme-api/theme-api.css">
                
            
        

    

    
        
    
        
    
        
    
        
    
        
    
        
    

        
    
    
    <meta name="HandheldFriendly" content="true"/>
    <meta name="viewport" content="width=device-width, initial-scale=1, user-scalable=no">
    <meta name="apple-mobile-web-app-capable" content="yes">
    <meta name="apple-mobile-web-app-status-bar-style" content="black">
    <link rel="apple-touch-icon-precomposed" sizes="152x152" href="../../gitbook/images/apple-touch-icon-precomposed-152.png">
    <link rel="shortcut icon" href="../../gitbook/images/favicon.ico" type="image/x-icon">

    
    <link rel="next" href="remove.html" />
    
    
    <link rel="prev" href="join.html" />
    

    </head>
    <body>
        
<div class="book">
    <div class="book-summary">
        
            
<div id="book-search-input" role="search">
    <input type="text" placeholder="Type to search" />
</div>

            
                <nav role="navigation">
                


<ul class="summary">
    
    

    

    
        
        <li class="header">Overview</li>
        
        
    
        <li class="chapter " data-level="1.1" data-path="../../">
            
                <a href="../../">
            
                    
                    Installation
            
                </a>
            

            
        </li>
    
        <li class="chapter " data-level="1.2" data-path="../../changelog.html">
            
                <a href="../../changelog.html">
            
                    
                    Change Log
            
                </a>
            

            
        </li>
    
        <li class="chapter " data-level="1.3" data-path="../../testing.html">
            
                <a href="../../testing.html">
            
                    
                    Testing
            
                </a>
            

            
        </li>
    
        <li class="chapter " data-level="1.4" data-path="../../contributions.html">
            
                <a href="../../contributions.html">
            
                    
                    Contributions
            
                </a>
            

            
        </li>
    

    
        
        <li class="header">Actions</li>
        
        
    
        <li class="chapter " data-level="2.1" data-path="../array.html">
            
                <a href="../array.html">
            
                    
                    Array
            
                </a>
            

            
            <ul class="articles">
                
    
        <li class="chapter " data-level="2.1.1" data-path="filter.html">
            
                <a href="filter.html">
            
                    
                    filter
            
                </a>
            

            
        </li>
    
        <li class="chapter " data-level="2.1.2" data-path="join.html">
            
                <a href="join.html">
            
                    
                    join
            
                </a>
            

            
        </li>
    
        <li class="chapter active" data-level="2.1.3" data-path="map.html">
            
                <a href="map.html">
            
                    
                    map
            
                </a>
            

            
        </li>
    
        <li class="chapter " data-level="2.1.4" data-path="remove.html">
            
                <a href="remove.html">
            
                    
                    remove
            
                </a>
            

            
        </li>
    
        <li class="chapter " data-level="2.1.5" data-path="sort.html">
            
                <a href="sort.html">
            
                    
                    sort
            
                </a>
            

            
        </li>
    
        <li class="chapter " data-level="2.1.6" data-path="swap.html">
            
                <a href="swap.html">
            
                    
                    swap
            
                </a>
            

            
        </li>
    

            </ul>
            
        </li>
    
        <li class="chapter " data-level="2.2" data-path="../object.html">
            
                <a href="../object.html">
            
                    
                    Objects
            
                </a>
            

            
            <ul class="articles">
                
    
        <li class="chapter " data-level="2.2.1" data-path="../object/clear.html">
            
                <a href="../object/clear.html">
            
                    
                    clear
            
                </a>
            

            
        </li>
    
        <li class="chapter " data-level="2.2.2" data-path="../object/rename.html">
            
                <a href="../object/rename.html">
            
                    
                    rename
            
                </a>
            

            
        </li>
    
        <li class="chapter " data-level="2.2.3" data-path="../object/update.html">
            
                <a href="../object/update.html">
            
                    
                    update
            
                </a>
            

            
        </li>
    
        <li class="chapter " data-level="2.2.4" data-path="../object/each.html">
            
                <a href="../object/each.html">
            
                    
                    each
            
                </a>
            

            
        </li>
    
        <li class="chapter " data-level="2.2.5" data-path="../object/toCamelCase.html">
            
                <a href="../object/toCamelCase.html">
            
                    
                    toCamelCase
            
                </a>
            

            
        </li>
    
        <li class="chapter " data-level="2.2.6" data-path="../object/toSnakeCase.html">
            
                <a href="../object/toSnakeCase.html">
            
                    
                    toSnakeCase
            
                </a>
            

            
        </li>
    

            </ul>
            
        </li>
    

    
        
        <li class="header">Recipes</li>
        
        
    
        <li class="chapter " data-level="3.1" data-path="../../recipes/shortcuts.html">
            
                <a href="../../recipes/shortcuts.html">
            
                    
                    Shortcuts
            
                </a>
            

            
        </li>
    
        <li class="chapter " data-level="3.2" data-path="../../recipes/dry.html">
            
                <a href="../../recipes/dry.html">
            
                    
                    DRY
            
                </a>
            

            
        </li>
    
        <li class="chapter " data-level="3.3" data-path="../../recipes/deep-objects.html">
            
                <a href="../../recipes/deep-objects.html">
            
                    
                    Deep Objects
            
                </a>
            

            
        </li>
    
        <li class="chapter " data-level="3.4" data-path="../../recipes/deep-arrays.html">
            
                <a href="../../recipes/deep-arrays.html">
            
                    
                    Deep Arrays
            
                </a>
            

            
        </li>
    

    

    <li class="divider"></li>

    <li>
        <a href="https://www.gitbook.com" target="blank" class="gitbook-link">
            Published with GitBook
        </a>
    </li>
</ul>


                </nav>
            
        
    </div>

    <div class="book-body">
        
            <div class="body-inner">
                
                    

<div class="book-header" role="navigation">
    

    <!-- Title -->
    <h1>
        <i class="fa fa-circle-o-notch fa-spin"></i>
        <a href="../.." >map</a>
    </h1>
</div>




                    <div class="page-wrapper" tabindex="-1" role="main">
                        <div class="page-inner">
                            
<div id="book-search-results">
    <div class="search-noresults">
    
                                <section class="normal markdown-section">
                                
                                <h2 id="map">Map</h2>
<p>[<a href="https://github.com/tuchk4/apop/blob/master/src/src/actions/array/map.js" target="_blank">source</a>]</p>
<blockquote>
<p>Adapter native <a href="https://www.w3schools.com/jsref/jsref_map.asp" target="_blank"><code>Array.map</code></a> for apop</p>
<h3 id="arguments">Arguments</h3>
<p>condition(Function): function(currentValue:element, index:Number, arr:Array)</p>
<h3 id="returns">Returns</h3>
<p>(Function): Returns Function that return new mapped Array.</p>
</blockquote>
<h3 id="examples">Examples:</h3>
<p><div class="codetabs"><div class="codetabs-header"><div class="tab active" data-codetab="0">Simple</div><div class="tab" data-codetab="1">Action</div><div class="tab" data-codetab="2">Recursive</div></div><div class="codetabs-body"><div class="tab active" data-codetab="0"><pre><code class="lang-js"><span class="hljs-keyword">let</span> ap = <span class="hljs-built_in">require</span>(<span class="hljs-string">&apos;apop/ap&apos;</span>);
ap.map(i =&gt; i + <span class="hljs-number">1</span>)([<span class="hljs-number">1</span>, <span class="hljs-number">2</span>, <span class="hljs-number">3</span>, <span class="hljs-number">4</span>, <span class="hljs-number">5</span>, [<span class="hljs-number">6</span>, <span class="hljs-number">7</span>, <span class="hljs-number">8</span>, <span class="hljs-number">9</span>, <span class="hljs-number">10</span>]])
<span class="hljs-comment">// =&gt; [2, 3, 4, 5, 6, &quot;6,7,8,9,101&quot; ]</span></code></pre></div><div class="tab" data-codetab="1"><pre><code class="lang-js"><span class="hljs-keyword">let</span> map = <span class="hljs-built_in">require</span>(<span class="hljs-string">&apos;apop/ap/map&apos;</span>);
map(i =&gt; i + <span class="hljs-number">1</span>)([<span class="hljs-number">1</span>, <span class="hljs-number">2</span>, <span class="hljs-number">3</span>, <span class="hljs-number">4</span>, <span class="hljs-number">5</span>, [<span class="hljs-number">6</span>, <span class="hljs-number">7</span>, <span class="hljs-number">8</span>, <span class="hljs-number">9</span>, <span class="hljs-number">10</span>]])
<span class="hljs-comment">// =&gt; [2, 3, 4, 5, 6, &quot;6,7,8,9,101&quot; ]</span></code></pre></div><div class="tab" data-codetab="2"><pre><code class="lang-js"><span class="hljs-keyword">let</span> ap = <span class="hljs-built_in">require</span>(<span class="hljs-string">&apos;apop/ap&apos;</span>);
ap.recursive(ap.map(i =&gt; i + <span class="hljs-number">1</span>))([<span class="hljs-number">1</span>, <span class="hljs-number">2</span>, <span class="hljs-number">3</span>, <span class="hljs-number">4</span>, <span class="hljs-number">5</span>, [<span class="hljs-number">6</span>, <span class="hljs-number">7</span>, <span class="hljs-number">8</span>, <span class="hljs-number">9</span>, <span class="hljs-number">10</span>]])
<span class="hljs-comment">// =&gt; [2, 3, 4, 5, 6, [7, 8, 9, 10, 11]]</span></code></pre></div></div></div></p>
<h2 id="interactive-example">Interactive Example:</h2>
<p><pre class="pg-runkit">
let ap = require(&apos;apop/ap&apos;);
let formula = ap(
    ap.filter(currentValue =&gt; currentValue &lt; 4), // =&gt; [3, 2, 1]
    ap.sort((a, b) =&gt; a - b), // =&gt; [1, 2, 3]
    ap.map(currentValue =&gt; ({id: currentValue})) // =&gt; [{id:1}, {id:2}, {id:3}]
);

formula([5, 4, 3, 2, 1])
// =&gt; [{id:1}, {id:2}, {id:3}]
</pre></p>
<script src="../../_gitbook_plugin_analytics.js"></script>
                                
                                </section>
                            
    </div>
    <div class="search-results">
        <div class="has-results">
            
            <h1 class="search-results-title"><span class='search-results-count'></span> results matching "<span class='search-query'></span>"</h1>
            <ul class="search-results-list"></ul>
            
        </div>
        <div class="no-results">
            
            <h1 class="search-results-title">No results matching "<span class='search-query'></span>"</h1>
            
        </div>
    </div>
</div>

                        </div>
                    </div>
                
            </div>

            

        
    </div>

    <script>
        var gitbook = gitbook || [];
        gitbook.push(function() {
            gitbook.page.hasChanged({"page":{"title":"map","level":"2.1.3","depth":2,"next":{"title":"remove","level":"2.1.4","depth":2,"path":"actions/array/remove.md","ref":"actions/array/remove.md","articles":[]},"previous":{"title":"join","level":"2.1.2","depth":2,"path":"actions/array/join.md","ref":"actions/array/join.md","articles":[]},"dir":"ltr"},"config":{"plugins":["theme-api","timeline","analytics","runkit","codetabs","copy-code-button","disqus"],"root":"./gitbook","styles":{"website":"styles/website.css","pdf":"styles/pdf.css","epub":"styles/epub.css","mobi":"styles/mobi.css","ebook":"styles/ebook.css","print":"styles/print.css"},"pluginsConfig":{"timeline":{},"disqus":{"useIdentifier":false,"shortName":"apop-1"},"analytics":{"google":"UA-12666422-9"},"search":{},"runkit":{},"codetabs":{},"lunr":{"maxIndexSize":1000000,"ignoreSpecialCharacters":false},"fontsettings":{"theme":"white","family":"sans","size":2},"highlight":{},"theme-api":{"languages":[{"lang":"js","name":"JavaScript","default":true}],"split":true,"theme":"light"},"copy-code-button":{},"sharing":{"facebook":true,"twitter":true,"google":false,"weibo":false,"instapaper":false,"vk":false,"all":["facebook","google","twitter","weibo","instapaper"]},"theme-default":{"styles":{"website":"styles/website.css","pdf":"styles/pdf.css","epub":"styles/epub.css","mobi":"styles/mobi.css","ebook":"styles/ebook.css","print":"styles/print.css"},"showLevel":false}},"theme":"default","pdf":{"pageNumbers":true,"fontSize":12,"fontFamily":"Arial","paperSize":"a4","chapterMark":"pagebreak","pageBreaksBefore":"/","margin":{"right":62,"left":62,"top":56,"bottom":56}},"structure":{"langs":"LANGS.md","readme":"README.md","glossary":"GLOSSARY.md","summary":"SUMMARY.md"},"variables":{"rep":"https://github.com/tuchk4/apop/blob/master/src"},"gitbook":">=3.0.0"},"file":{"path":"actions/array/map.md","mtime":"2017-07-25T14:14:39.106Z","type":"markdown"},"gitbook":{"version":"3.2.2","time":"2017-07-25T14:43:05.023Z"},"basePath":"../..","book":{"language":""}});
        });
    </script>
</div>

        
    <script src="../../gitbook/gitbook.js"></script>
    <script src="../../gitbook/theme.js"></script>
    
        
        <script src="https://embed.runkit.com"></script>
        
    
        
        <script src="../../gitbook/gitbook-plugin-runkit/plugin.js"></script>
        
    
        
        <script src="../../gitbook/gitbook-plugin-codetabs/codetabs.js"></script>
        
    
        
        <script src="../../gitbook/gitbook-plugin-copy-code-button/toggle.js"></script>
        
    
        
        <script src="https://cdnjs.cloudflare.com/ajax/libs/URI.js/1.16.1/URI.min.js"></script>
        
    
        
        <script src="../../gitbook/gitbook-plugin-disqus/plugin.js"></script>
        
    
        
        <script src="../../gitbook/gitbook-plugin-search/search-engine.js"></script>
        
    
        
        <script src="../../gitbook/gitbook-plugin-search/search.js"></script>
        
    
        
        <script src="../../gitbook/gitbook-plugin-lunr/lunr.min.js"></script>
        
    
        
        <script src="../../gitbook/gitbook-plugin-lunr/search-lunr.js"></script>
        
    
        
        <script src="../../gitbook/gitbook-plugin-sharing/buttons.js"></script>
        
    
        
        <script src="../../gitbook/gitbook-plugin-fontsettings/fontsettings.js"></script>
        
    
        
        <script src="../../gitbook/gitbook-plugin-theme-api/theme-api.js"></script>
        
    

    </body>
</html>

