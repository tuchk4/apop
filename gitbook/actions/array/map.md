## Map
[[source]({{book.rep}}/src/actions/array/map.js)]

> Adapter native [`Array.map`](https://www.w3schools.com/jsref/jsref_map.asp) for apop
> ### Arguments
> 
> condition(Function): function(currentValue:element, index:Number, arr:Array)
> 
> ### Returns
> 
> (Function): Returns Function that return new mapped Array.


### Examples: 
{% codetabs name="Simple", type="js" -%} 
let ap = require('apop/ap');
ap.map(i => i + 1)([1, 2, 3, 4, 5, [6, 7, 8, 9, 10]])
// => [2, 3, 4, 5, 6, "6,7,8,9,101" ]

{%- language name="Action", type="js" -%}
let map = require('apop/ap/map');
map(i => i + 1)([1, 2, 3, 4, 5, [6, 7, 8, 9, 10]])
// => [2, 3, 4, 5, 6, "6,7,8,9,101" ]

{%- language name="Recursive", type="js" -%}
let ap = require('apop/ap');
ap.recursive(ap.map(i => i + 1))([1, 2, 3, 4, 5, [6, 7, 8, 9, 10]])
// => [2, 3, 4, 5, 6, [7, 8, 9, 10, 11]]

{%- endcodetabs %}


## Interactive Example:

{% runkit %}
let ap = require('apop/ap');
let formula = ap(
    ap.filter(currentValue => currentValue < 4), // => [3, 2, 1]
    ap.sort((a, b) => a - b), // => [1, 2, 3]
    ap.map(currentValue => ({id: currentValue})) // => [{id:1}, {id:2}, {id:3}]
);

formula([5, 4, 3, 2, 1])
// => [{id:1}, {id:2}, {id:3}]
{% endrunkit %}


