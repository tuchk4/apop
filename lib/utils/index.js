/** @namespace utils */
import parseArray from '~/utils/parseArray';
import parseObject from '~/utils/parseObject';
import reduce from '~/utils/reduce';

module.exports =  {parseArray, parseObject, reduce};