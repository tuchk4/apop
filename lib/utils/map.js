/**
 * Convert Map to Object
 * @memberof utils
 * @param {Map} origin Origin array
 *
 * @returns {Object} Transformed array of object by actions
 */
export function decode(origin) {
  let obj = Object.create(null);
  for (let [k, v] of origin) {
    // We donâ€™t escape the key '__proto__'
    // which can cause problems on older engines
    obj[k] = v;
  }
  return obj;
}
/**
 * Convert Object to Map
 * @memberof utils
 * @param {Object} origin Origin array
 *
 * @returns {Map} Transformed array of object by actions
 */
export function encode(origin) {
  let strMap = new Map();
  for (let k of Object.keys(origin)) {
    strMap.set(k, origin[k]);
  }
  return strMap;
}
export default { decode, encode };
