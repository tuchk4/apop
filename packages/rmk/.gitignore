.DS_Store
/.git
/.idea
/npm-debug.log
/node_modules
/dist
recursive
*.json
/*.js
/*.map
!__tests__
!src
!rollup-build.js
