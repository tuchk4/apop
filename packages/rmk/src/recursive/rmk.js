import rmkActions from './../actions';
import rmkParse from './../utils/parse';

const recursiveActions = {};

Object.keys(rmkActions).forEach(action => {
  recursiveActions[action] = config =>
    rmkParse({
      actions: [rmkActions[action](config)],
      isRecursive: true,
    });
});

const recursive = Object.assign(
  (...actions) => rmkParse({ actions, isRecursive: true }),
  recursiveActions
);

export default recursive;
