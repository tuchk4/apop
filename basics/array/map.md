# Array Action - Map

Adapter for native [`array.map(function(currentValue, index, arr))`](/ https://www.w3schools.com/jsref/jsref_map.asp)

### Imports for simple Array parsing

Import ap with actions:

```js
import ap from 'apop/ap';
ap.map(conditions)(data)
```

Import filter action:

```js
import map from 'apop/ap/map'
map(conditions)(data)
```

### Import for recursive Array parsing

Import recursive ap:

```js
import recursive from 'apop/ap/recursive'
recursive.map(conditions)(data)
```

Import direct recursive action:

```js
import join from 'apop/ap/recursive/join'
map(conditions)(data)
```

Import recursive ap and filter action:

```js
import join from 'apop/ap/join'
import recursive from 'apop/ap/recursive'

recursive(map(conditions))(data)
```

## Examples

Equal to array.map

```js
let formula = ap.map(currentValue => currentValue + 1);

formula([1, 2, 3, 4, 5])
// => [2, 3, 4, 5, 6]
```

Map array of objects

```js
let formula = ap.map((currentValue, index) => {
    return {id: currentValue.id + index};
});

formula([{id: 1}, {id: 2}, {id: 3}, {id: 4}, {id: 5}]);
// => [{id: 1}, {id: 3}, {id: 5}, {id: 7}, {id: 9}]
```

Usage in flow

```js
let formula = ap(
    ap.filter(currentValue => currentValue > 2), // => [3, 2, 1]
    ap.sort((a, b) => a - b) // => [1, 2, 3]
    ap.map(currentValue => {id: currentValue})
);

formula([5, 4, 3, 2, 1])
// => [{id:1}, {id:2}, {id:3}]
```



