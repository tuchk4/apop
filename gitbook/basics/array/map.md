## Map
[[source](../../../src/actions/array/map.js)]

> Adapter native [`Array.map`](https://www.w3schools.com/jsref/jsref_map.asp) for apop
> ### Arguments
> 
> condition (Function): function(currentValue, index, arr)
> 
> ### Returns
> 
> (Function): Returns Function that return new mapped Array.


### Examples: 
{% codetabs name="AP", type="js" -%} 
import ap from 'apop/ap';
ap.map(i => i + 1)([1, 2, 3, 4, 5, [6, 7, 8, 9, 10, [11, 12, 13, 14, 15]]])
// => [2, 3, 4, 5, 6, "6,7,8,9,10,11,12,13,14,151" ]

{%- language name="Action", type="js" -%}
import map from 'apop/ap/map'
map(i => i + 1)([1, 2, 3, 4, 5, [6, 7, 8, 9, 10, [11, 12, 13, 14, 15]]])
// => [2, 3, 4, 5, 6, "6,7,8,9,10,11,12,13,14,151" ]

{%- language name="Recursive", type="js" -%}
import recursive from 'apop/ap/recursive'
recursive.map(i => i + 1)([1, 2, 3, 4, 5, [6, 7, 8, 9, 10, [11, 12, 13, 14, 15]]])
// => [2, 3, 4, 5, 6, [7, 8, 9, 10, 11, [12, 13, 14, 15, 16]]]

{%- language name="Recursive Action", type="js" -%}
import map from 'apop/ap/recursive/map'
map(i => i + 1)([1, 2, 3, 4, 5, [6, 7, 8, 9, 10, [11, 12, 13, 14, 15]]])
// => [2, 3, 4, 5, 6, [7, 8, 9, 10, 11, [12, 13, 14, 15, 16]]]
{%- endcodetabs %}


## Interactive Example:

{% runkit %}
let ap = require('apop/ap');
let formula = ap(
    ap.filter(currentValue => currentValue < 4), // => [3, 2, 1]
    ap.sort((a, b) => a - b), // => [1, 2, 3]
    ap.map(currentValue => {id: currentValue}) // => [{id:1}, {id:2}, {id:3}]
);

formula([5, 4, 3, 2, 1])
// => [{id:1}, {id:2}, {id:3}]
{% endrunkit %}


