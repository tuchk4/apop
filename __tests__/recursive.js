import isArray from 'lodash/isArray';
import rmk from '../lib';
import * as mapUtils from '../lib/utils/map';

const sample = {
  foo_bar: 0,
  remove_baz: null,
  array_key: [1, 2, 3],
  array_objects: [{ a_b: 'a', b_c: 'b' }],
};

const data = {
  asObject: {
    ...sample,
    deep_obj: sample,
  },
  asArray: [
    {
      children: [],
      name: '1',
      url: '/1',
    },
    {
      children: [],
      name: '2',
      url: '/v2/2',
    },
    {
      children: [
        {
          children: [],
          name: '3',
          url: '/v2/3',
        },
        {
          children: [
            {
              children: [],
              name: '3',
              url: '/3',
            },
            {
              children: [],
              name: '4',
              url: '/v2/4',
            },
            {
              children: [],
              name: '5',
              url: '/v2/5',
            },
          ],
          name: '6',
        },
      ],
    },
  ],
};

const VERSIONS = {
  V1: 'v1',
  V2: 'v2',
};

const versionLinkFn = (url, isLink) => {
  let version;
  if (url && isLink) {
    version = url.startsWith(`/${VERSIONS.V2}`) ? VERSIONS.V2 : VERSIONS.V1;
  }

  return version;
};

const normalizeUrl = (url, version) =>
  version === VERSIONS.V2 ? url.replace(`/${VERSIONS.V2}`, '') : url;
const normalizeName = (name, version) =>
  version === VERSIONS.V1 ? `${name} old` : name;

test('Object recursive', () => {
  const formula = rmk(
    rmk.recursive(
      [
        rmk.toCamelCase(),
        rmk.update({
          arrayKeyStr: localState =>
            isArray(localState.arrayKey) ? localState.arrayKey.join(',') : null,
        }),
        rmk.rename({
          arrayKeyStr: 'renamedStr',
        }),
        rmk.clear(),
      ],
      3
    )
  );

  const result = formula(data.asObject);

  expect(result).toMatchSnapshot();
});

test('Array recursive', () => {
  const transformMainMenu = rmk(
    rmk.recursive(
      [
        rmk.clear(),
        rmk.toCamelCase(),
        rmk.rename({
          children: 'nodes',
        }),
      ],
      8
    ),
    rmk.recursive(
      [
        rmk.update({
          isCategory: localState =>
            Object.prototype.hasOwnProperty.call(localState, 'nodes'),
          isLink: localState =>
            !Object.prototype.hasOwnProperty.call(localState, 'nodes'),
        }),
      ],
      8
    ),
    rmk.recursive(
      [
        rmk.update({
          version: localState =>
            versionLinkFn(localState.url, localState.isLink),
        }),
      ],
      8
    ),
    rmk.recursive(
      [
        rmk.update({
          url: localState => normalizeUrl(localState.url, localState.version),
          name: localState =>
            normalizeName(localState.name, localState.version),
          isOpen: localState => {
            let isOpen = null;
            if (localState.isCategory) {
              isOpen = false;
            }
            return isOpen;
          },
        }),
        rmk.clear(),
      ],
      8
    )
  );
  const result = transformMainMenu(data.asArray);

  expect(result).toMatchSnapshot();
});

test('Array recursive with map', () => {
  const data = new Set();

  data.add(new Map([['x', '0'], ['y', null]]));
  data.add(new Map([['x', '1'], ['y', '1']]));
  data.add(new Map([['x', null], ['y', '2']]));
  data.add(new Map([['x', '3'], ['y', '3']]));
  data.add(new Map([['x', '4'], ['y', '4']]));
  data.add(new Map([['deep', data]]));

  const formula = rmk(rmk.recursive([rmk.clear()], 6));
  const result = formula(data);
  expect(result).toMatchSnapshot();
});
