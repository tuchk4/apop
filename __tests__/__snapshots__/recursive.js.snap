// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Array recursive 1`] = `
Array [
  Object {
    "isCategory": false,
    "isLink": true,
    "name": "1 old",
    "url": "/1",
    "version": "v1",
  },
  Object {
    "isCategory": false,
    "isLink": true,
    "name": "2",
    "url": "/2",
    "version": "v2",
  },
  Object {
    "isCategory": true,
    "isLink": false,
    "isOpen": false,
    "nodes": Array [
      Object {
        "isCategory": false,
        "isLink": true,
        "name": "3",
        "url": "/3",
        "version": "v2",
      },
      Object {
        "isCategory": true,
        "isLink": false,
        "isOpen": false,
        "name": "6",
        "nodes": Array [
          Object {
            "isCategory": false,
            "isLink": true,
            "name": "3 old",
            "url": "/3",
            "version": "v1",
          },
          Object {
            "isCategory": false,
            "isLink": true,
            "name": "4",
            "url": "/4",
            "version": "v2",
          },
          Object {
            "isCategory": false,
            "isLink": true,
            "name": "5",
            "url": "/5",
            "version": "v2",
          },
        ],
      },
    ],
  },
]
`;

exports[`Array recursive with map 1`] = `
Array [
  Object {
    "x": "0",
  },
  Object {
    "x": "1",
    "y": "1",
  },
  Object {
    "y": "2",
  },
  Object {
    "x": "3",
    "y": "3",
  },
  Object {
    "x": "4",
    "y": "4",
  },
  Object {
    "deep": Array [
      Object {
        "x": "0",
      },
      Object {
        "x": "1",
        "y": "1",
      },
      Object {
        "y": "2",
      },
      Object {
        "x": "3",
        "y": "3",
      },
      Object {
        "x": "4",
        "y": "4",
      },
      Object {
        "deep": Array [
          Object {
            "x": "0",
          },
          Object {
            "x": "1",
            "y": "1",
          },
          Object {
            "y": "2",
          },
          Object {
            "x": "3",
            "y": "3",
          },
          Object {
            "x": "4",
            "y": "4",
          },
          Object {
            "deep": Array [
              null,
              null,
              null,
              null,
              null,
              null,
            ],
          },
        ],
      },
    ],
  },
]
`;

exports[`Object recursive 1`] = `
Object {
  "arrayKey": Array [
    1,
    2,
    3,
  ],
  "arrayObjects": Array [
    Object {
      "aB": "a",
      "bC": "b",
    },
  ],
  "deepObj": Object {
    "arrayKey": Array [
      1,
      2,
      3,
    ],
    "fooBar": 0,
    "renamedStr": "1,2,3",
  },
  "fooBar": 0,
  "renamedStr": "1,2,3",
}
`;
